### PR TITLE
[Feature] Verify and Support for DeepSeek-OCR-2

### PR DESCRIPTION
**Pull request by liutianyang-2026**
### What this PR does / why we need it?

This PR adds and validates **DeepSeek-OCR-2** support in vllm-ascend, and includes compatibility/stability fixes found during bring-up.

Main changes:

- Initialize and verify DeepSeek-OCR-2 support flow on Ascend.
- Fix graph compilation failure in `QKNormRopeFusionPass` when `rope_dim <= 0`.
  - Root cause: DeepSeek-OCR-2 sets `max_position_embeddings=0`, which can lead to `rope_dim=0`.
- Add a compatibility patch for newer `transformers`, and fix lm-eval failures on DeepSeek models.
  - Context: DeepSeek-OCR-2 depends on `LlamaFlashAttention2` (though as a fallback), which is deprecated in later `transformers` versions.
- Add DeepSeek-OCR-2 tutorial/documentation.
- Add e2e model config for correctness validation:
  - `tests/e2e/models/configs/DeepSeek-OCR-2.yaml`
- Apply CI/lint fixes for all related changes.

Why needed:

- Enable reliable DeepSeek-OCR-2 usage on Ascend.
- Avoid graph compile errors and lm-eval compatibility issues on relevant DeepSeek code paths.

### Does this PR introduce _any_ user-facing change?

Yes.

- Adds DeepSeek-OCR-2 support and usage documentation.
- Improves runtime compatibility/stability for affected DeepSeek model scenarios (newer `transformers`, lm-eval path, graph compilation edge case).

### How was this patch tested?

- Added e2e model config:
  - `tests/e2e/models/configs/DeepSeek-OCR-2.yaml`
- Ran CI/lint alignment in follow-up commit (`apply ci & lint`).

- vLLM version: v0.18.0
- vLLM main: https://github.com/vllm-project/vllm/commit/35141a7eeda941a60ad5a4956670c60fd5a77029

### CODE REVIEW

**Comment by liutianyang-2026** [comment]
## Local Test Results
- e2e pytest:
  [log.log](https://github.com/user-attachments/files/26259252/log.log)截屏2026-03-25 22 21 07截屏2026-03-25 22 21 40
- server test:截屏2026-03-25 22 24 47
- performance test:截屏2026-03-25 22 29 17
- CI/Lint test:
   截屏2026-03-26 11 02 58

**Comment by liutianyang-2026** [comment]
# SKILL: Adapt DeepSeek-OCR-2 (VLM) for vllm-ascend

## Skill Overview
- **Skill Name**: Adapt DeepSeek-OCR-2 (VLM) for vllm-ascend
- **Core Objective**: Complete the end-to-end verification, dependency patching, compilation pass optimization, and CI evaluation integration for a multi-modal Large Language Model featuring complex `trust_remote_code` on Ascend NPU.
- **Related Domains**: Multiprocessing Context Isolation, Framework-level Monkey Patching, VLM Baseline Evaluation, Graph Compilation Passes.

## Workflows & Best Practices

### 1. Process Isolation for Remote Code Loading
- **The Pitfall**: When `trust_remote_code=True` is enabled, loading custom HuggingFace code in the main process prematurely triggers PyTorch's C++ initialization. When vLLM subsequently spawns child processes (EngineCore) using the default `fork` method, the children inherit a polluted OpenMP thread pool state on ARM/Ascend architectures, leading to random `Invalid thread pool` crashes.
- **Best Practice**: You must enforce a `spawn` multiprocessing method and strict OpenMP thread limits via environment variables **before** any heavy libraries (like `vllm` or `torch`) are imported.
  ```bash
  export VLLM_WORKER_MULTIPROC_METHOD=spawn
  export OMP_WAIT_POLICY=PASSIVE
  export OMP_NUM_THREADS=1
  ```

### 2. Upstream Dependency Patching via Platform Hooks
- **The Pitfall**: Legacy remote code (e.g., DeepSeek's `modeling_deepseekv2.py`) may import classes that have been removed in newer versions of the `transformers` library (like `LlamaFlashAttention2`), causing an `ImportError` during `AutoConfig` loading.
- **Best Practice**: Never modify the user's HuggingFace cache or the model's source code. Instead, utilize `vllm-ascend`'s dynamic patch mechanism. Inject a dummy alias (e.g., mapping `LlamaFlashAttention2` to `LlamaAttention`) in `vllm_ascend/patch/platform/` to bypass the inheritance check during the earliest stage of the framework lifecycle.

### 3. Handling Anomalous Model Configurations in Graph Compilation
- **The Pitfall**: DeepSeek-OCR-2 sets `max_position_embeddings=0` in its configuration. This anomalous value cascades into the graph compilation phase, specifically causing `rope_dim <= 0`, which leads to a compilation failure in the `QKNormRopeFusionPass` (a pass designed to optimize QK-Norm and RoPE execution on NPU).
- **Best Practice**: When adapting new models, always inspect their specific configuration anomalies (like zeroed-out position embeddings for pure vision/OCR tasks). Update the graph compilation passes to safely skip or handle invalid dimensions (e.g., `if rope_dim <= 0: return`) to prevent the entire compilation pipeline from breaking.

### 4. Efficient Client-Server Debugging & VLM Evaluation
- **The Pitfall**: Mixing the evaluation harness (`lm_eval`) and the vLLM engine in a single process requires reloading tens of gigabytes of model weights for every configuration tweak, drastically slowing down debugging. Furthermore, `lm_eval`'s dataset downloads can be unexpectedly hijacked by `VLLM_USE_MODELSCOPE`.
- **Best Practice**: 
  - **Decoupled Architecture**: Launch a persistent API Server (`vllm.entrypoints.openai.api_server`) in one terminal. Use `lm_eval --model local-chat-completions` as a lightweight client in another terminal to quickly iterate on VLM datasets like `doc_vqa`.
  - **Baseline Establishment**: Since base OCR models lack Chat Templates and tend to output verbose descriptions on zero-shot prompts, employ Contextual Learning (e.g., `--num_fewshot 3`) to regularize the output format. Accept a non-zero, deterministic score (even if relatively low) as a valid regression baseline for hardware execution correctness.

## Effective Prompts for AI Collaboration
- *"The model runs correctly occasionally but crashes with 'Invalid thread pool' randomly. Could you analyze the C++ stack trace to determine if there's a race condition related to the `fork` mechanism when loading Remote Code?"*
- *"If modifying the HuggingFace cache is prohibited, at which lifecycle stage of vllm-ascend (Worker or Platform) should I inject a dependency fix patch to intercept `AutoConfig`?"*

**Comment by liutianyang-2026** [comment]
@wjunLu

**Comment by wjunLu** [comment]
Good job!

Please  take a look @MengqingCao

File: tests/e2e/models/configs/DeepSeek-OCR-2.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+model_name: "deepseek-ai/DeepSeek-OCR-2"
```
Comment (zhangxinyuehfad) [inline tests/e2e/models/configs/DeepSeek-OCR-2.yaml:1]: plz add new mode into https://github.com/vllm-project/vllm-ascend/blob/main/.github/workflows/misc/model_list.json, and we will add lable to download

Comment (zhangxinyuehfad) [inline tests/e2e/models/configs/DeepSeek-OCR-2.yaml:1]: plz add new mode into https://github.com/vllm-project/vllm-ascend/blob/main/tests/e2e/models/configs/accuracy_groups_a2.json.

**Comment by jjsnam** [comment]
@zhangxinyuehfad all fixed

File: tests/e2e/models/configs/accuracy_groups_a2.json [hunk 2]
```diff
@@ -47,7 +47,8 @@
       "os": "linux-aarch64-a2b3-1",
       "model_list": [
         "gemma-3-4b-it",
-        "internlm3-8b-instruct"
+        "internlm3-8b-instruct",
+        "deepseek-ai/DeepSeek-OCR-2"
```
Comment (zhangxinyuehfad) [inline tests/e2e/models/configs/accuracy_groups_a2.json]: change to `DeepSeek-OCR-2`

**Comment by MrZ20** [comment]
Community lint rule changes, before merging please rebase and add the md file path to the exclude list of `pyproject.toml::[tool.check_docs_yaml_sync]`.
